File: 2025/Day-01/Part02.py
import os

def read_input(filename: str) -> list[str]:
    data: list[str] = []
    with open(filename, 'r') as f:
        data = [line.strip() for line in f.readlines()]
    return data

def dial(position: int, instruction: str, dial_size: int = 100) -> tuple[int, int]:
    direction: str = instruction[0]
    distance: int = int(instruction[1:])
    hits: int = 0
    new_position: int = 0
    
    if direction == "R":
        # Count how many times we cross or land on 0 going right
        # From position, we need to reach 0, then every dial_size after that
        if position == 0:
            # Starting at 0, we hit it again after every full loop
            hits = distance // dial_size
        else:
            # Distance to first 0 is (dial_size - position)
            # Then we hit 0 every dial_size after that
            remaining = distance - (dial_size - position)
            if distance >= (dial_size - position):
                hits = 1 + max(0, remaining // dial_size)
        new_position = (position + distance) % dial_size
    elif direction == "L":
        # Count how many times we cross or land on 0 going left
        if position == 0:
            # Starting at 0, we hit it again after every full loop
            hits = distance // dial_size
        else:
            # Distance to first 0 is position
            # Then we hit 0 every dial_size after that
            remaining = distance - position
            if distance >= position:
                hits = 1 + max(0, remaining // dial_size)
        new_position = (position - distance) % dial_size
    
    return new_position, hits

def main(data: list[str]) -> int:
    position: int = 50
    hits: int = 0
    
    for instruction in data:
        position, hits_on_dial = dial(position, instruction)
        if hits_on_dial:
            hits += hits_on_dial
    
    return hits

if __name__ == "__main__":
    input_file: str = os.path.dirname(os.path.abspath(__file__)) + r"\\input.txt"
    data: list[str] = read_input(input_file)
    
    print(f"Part 2: {main(data)}")

File: 2025/Day-01-dart/part02.dart
import 'dart:io';

/// Read and parse input file
List<String> readInput(String filename) {
  return File(filename).readAsLinesSync();
}

(int, int) dial(String instruction, int position, {int dialSize = 100}) {
  String direction = instruction[0];
  int distance = int.parse(instruction.substring(1));
  int hits = 0;
  int new_position = 0;

  if (direction == "R") {
    for (int i = 1; i <= distance; i++) {
      new_position = (position + i) % dialSize;
      if (new_position == 0) {
        hits += 1;
      }
    }
    position = new_position;
  } else if (direction == "L") {
    for (int i = 1; i <= distance; i++) {
      new_position = (position - i) % dialSize;
      if (new_position == 0) {
        hits += 1;
      }
    }
    position = new_position;
  }

  return (position, hits);
}

int solve(List<String> data) {
  int position = 50;
  int hits = 0;

  for (String instruction in data) {
    int hits_on_dial;
    (position, hits_on_dial) = dial(instruction, position);
    hits += hits_on_dial;
  }
  return hits;
}

void main() {
  final scriptDir = File(Platform.script.toFilePath()).parent.path;
  final inputFile = '$scriptDir${Platform.pathSeparator}input.txt';
  final data = readInput(inputFile);

  print("Part 2: ${solve(data)}");
}
